server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,287 @@
+// org.gnome.keyring.Prompter
+// https://gitlab.gnome.org/GNOME/gcr/-/blob/main/gcr/org.gnome.keyring.Prompter.xml
+
+use clap::error::Result;
+use oo7::dbus::ServiceError;
+use serde::{Deserialize, Serialize};
+use zbus::{
+    interface, proxy,
+    zvariant::{DeserializeDict, OwnedObjectPath, OwnedValue, SerializeDict, Type, Value},
+};
+
+use crate::{
+    gnome::secret_exchange::SecretExchange,
+    prompt::{Prompt, PromptRole},
+    service::Service,
+};
+
+#[derive(Debug, DeserializeDict, SerializeDict, Type)]
+#[zvariant(signature = "a{sv}")]
+// System prompt properties
+pub struct Properties {
+    title: Option<String>,
+    #[zvariant(rename = "choice-label")]
+    choice_label: Option<String>,
+    description: Option<String>,
+    message: Option<String>,
+    #[zvariant(rename = "caller-window")]
+    caller_window: Option<String>,
+    warning: Option<String>,
+    #[zvariant(rename = "password-new")]
+    password_new: Option<bool>,
+    #[zvariant(rename = "password-strength")]
+    password_strength: Option<u32>,
+    #[zvariant(rename = "choice-chosen")]
+    choice_chosen: Option<bool>,
+    #[zvariant(rename = "continue-label")]
+    continue_label: Option<String>,
+    #[zvariant(rename = "cancel-label")]
+    cancel_label: Option<String>,
+}
+
+#[derive(Debug, Type)]
+#[zvariant(signature = "s")]
+// Possible values for PromptReady reply parameter
+pub enum Reply {
+    No,
+    Yes,
+    Empty,
+}
+
+const NO: &str = "no";
+const YES: &str = "yes";
+const EMPTY: &str = "";
+
+impl Serialize for Reply {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::No => str::serialize(NO, serializer),
+            Self::Yes => str::serialize(YES, serializer),
+            Self::Empty => str::serialize(EMPTY, serializer),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Reply {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        match String::deserialize(deserializer)?.as_str() {
+            NO => Ok(Self::No),
+            YES => Ok(Self::Yes),
+            EMPTY => Ok(Self::Empty),
+            err => Err(serde::de::Error::custom(format!("Invalid reply {err}"))),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Type)]
+#[serde(rename_all = "lowercase")]
+#[zvariant(signature = "s")]
+// Possible values for PerformPrompt type parameter
```

Comment:
Should this implement the serialisation traits or the derived ones are correct?

Reply:
Yes, the derived ones are correct. I had to implement the serialisation and de-serialisation for `Reply` because of the Empty variant.